### ISSUE TITLE
VAE example

### ISSUE THREAD

**Issue by dmccloskey** [dmccloskey/EvoNet#51]
### Description
MNIST VAE example

### Objectives
- [x] Model cache for input and output nodes when adding nodes to the model
- [x] update checkCompleteModel to use input/output node cache
- [x] Update modelTrainer to allow for training with multiple lossFunction/output node pairs
- [x] Update model calculateError function to += error_ instead of assign
- [x] Update modelTrainer to include getters/setters for loss_functions and output_node pairs
- [x] Test modelTrainer
- [x] Update population trainer to remove output nodes and instead use modelTrainer
- [x] KL divergence loss function
- [x] addVAESampler to ModelBuilder
- [x] Gaussian sampler for input to VAE
- [x] BCEWithLogits
- Perceptual loss function (?)

### Training/Testing
- during training, set the sampler to sample from a Gaussian
- during testing, set the sampler to 0

### References (Vanilla VAE)
- https://towardsdatascience.com/intuitively-understanding-variational-autoencoders-1bfe67eb5daf
- https://towardsdatascience.com/teaching-a-variational-autoencoder-vae-to-draw-mnist-characters-978675c95776
- https://github.com/FelixMohr/Deep-learning-with-Python/blob/master/VAE.ipynb
- https://github.com/hwalsuklee/tensorflow-mnist-VAE
- https://arxiv.org/pdf/1312.6114.pdf
- https://github.com/pytorch/examples/blob/master/vae/main.py

### References (Perceptual loss)
- DOI 10.1109/WACV.2017.131
- http://krasserm.github.io/2018/07/27/dfc-vae/

### Reference (BCEWithLogits)
- https://pytorch.org/docs/stable/nn.html#bcewithlogitsloss
- https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits
- https://github.com/pytorch/pytorch/blob/d5a7e304fa858136eeffa1ebe35e22ed886bf256/torch/nn/functional.py#L859
- https://pytorch.org/docs/master/torch.html#torch.clamp
